e2e/pushpull_test.go

```diff
@@ -0,0 +1,147 @@
+	// - the registry must be reachable from the dind daemon on the same address/port
+	// Solution found is: fix the port of the registry to be the same internally and externally (fixed at 5000, could use something random)
+	// run the dind container in the same network namespace: this way 127.0.0.1:5000 both resolves to the registry from the client and from dind
+
+	swarm := NewContainer("docker:18.09-dind", 2375, "--insecure-registry", fmt.Sprintf("127.0.0.1:%d", registryPort))
+	swarm.Start(t, "--expose", strconv.FormatInt(int64(registryPort), 10),
+		"-p", fmt.Sprintf("%d:%d", registryPort, registryPort),
+		"-p", "2375")
+	defer swarm.Stop(t)
+
+	registry := NewContainer("registry:2", registryPort)
+	registry.StartWithContainerNetwork(t, swarm, "-e", "REGISTRY_VALIDATION_MANIFESTS_URLS_ALLOW=[^http]",
+		"-e", fmt.Sprintf("REGISTRY_HTTP_ADDR=0.0.0.0:%d", registryPort))
+	defer registry.StopNoFail()
+
+	// We  need two contexts:
+	// - one for `docker` so that it connects to the dind swarm created before
+	// - the target context for the invocation image to install within the swarm
+	cmd.Command = []string{dockerCli, "context", "create", "swarm-context", "--docker", fmt.Sprintf(`"host=tcp://%s"`, swarm.GetAddress(t)), "--default-stack-orchestrator", "swarm"}
+	icmd.RunCmd(cmd).Assert(t, icmd.Success)
+
+	// When creating a context on a Windows host we cannot use
+	// the unix socket but it's needed inside the invocation image.
+	// The workaround is to create a context with an empty host.
+	// This host will default to the unix socket inside the
+	// invocation image
+	cmd.Command = []string{dockerCli, "context", "create", "swarm-target-context", "--docker", "host=", "--default-stack-orchestrator", "swarm"}
+	icmd.RunCmd(cmd).Assert(t, icmd.Success)
+
+	// Initialize the swarm
+	cmd.Env = append(cmd.Env, "DOCKER_CONTEXT=swarm-context")
+	cmd.Command = []string{dockerCli, "swarm", "init"}
+	icmd.RunCmd(cmd).Assert(t, icmd.Success)
+	// Load the needed base cnab image into the swarm docker engine
+	cmd.Command = []string{dockerCli, "load", "-i", tmpDir.Join("cnab-app-base.tar.gz")}
+	icmd.RunCmd(cmd).Assert(t, icmd.Success)
+
+	info := dindSwarmAndRegistryInfo{
+		configuredCmd:   cmd,
+		registryAddress: registry.GetAddress(t),
+		swarmAddress:    swarm.GetAddress(t),
+		stopRegistry:    registry.StopNoFail,
+	}
+	todo(info)
+
+}
+
+func TestPushInstall(t *testing.T) {
+	runWithDindSwarmAndRegistry(t, func(info dindSwarmAndRegistryInfo) {
+		cmd := info.configuredCmd
+		ref := info.registryAddress + "/test/push-pull"
+		cmd.Command = []string{dockerApp, "push", "-t", ref, "--insecure-registries=" + info.registryAddress, filepath.Join("testdata", "push-pull", "push-pull.dockerapp")}
+		icmd.RunCmd(cmd).Assert(t, icmd.Success)
+
+		cmd.Command = []string{dockerApp, "install", "--insecure-registries=" + info.registryAddress, ref, "--name", t.Name()}
```

Comment:
This (and other tests) are going to populate `~/.docker/app-bundle-store` with things it finds (and some future commands might also manage to use the cache there).

I think we need a tempdir and `["--config", tmpdir]` type arrangement here.

Except `--config` doesn't actually seem to work -- I will figure that out now.

We should also in a few places validate that the `tmpdir` is getting populated (haven't read far enough through this PR to see if we do that yet)

Reply:
Can we honor the DOCKER_CONFIG env var for the app-bundle-store ? (it is set in this test)

Reply:
Yes, that seems to already be working, phew!

I should still figure out `--config` because it *should* work, but not critical to the correctness of this e2e test.